src/opera/api/controllers/security_controller.py

```diff
@@ -0,0 +1,177 @@
+import functools
+import re
+from base64 import b64encode
+
+import connexion
+import requests
+
+from opera.api.cli import CSAR_db, SQL_database
+from opera.api.openapi.models.just_message import JustMessage
+from opera.api.settings import Settings
+
+# use connection pool for OAuth tokeninfo
+adapter = requests.adapters.HTTPAdapter(pool_connections=100, pool_maxsize=100)
+session = requests.Session()
+for protocol in Settings.introspection_protocols:
+    session.mount(protocol, adapter)
+
+role_regex = r"(?P<domain>\w+)_(?P<type>\w+)_(?P<permissions>\w)"
+AADM_SUFFIX = "aadm"
+RM_SUFFIX = "rm"
+
+
+def check_api_key(apikey, required_scopes=None):
+    if not Settings.apiKey or apikey != Settings.apiKey:
+        return None
+
+    return {'scope': ['apiKey']}
+
+
+def token_info(access_token) -> dict:
+    request = {'token': access_token}
+    headers = {'Content-type': 'application/x-www-form-urlencoded'}
+    token_info_url = Settings.oidc_introspection_endpoint_uri
+    # TODO add multiple client support
+    basic_auth_string = '{0}:{1}'.format(
+        Settings.oidc_client_id,
+        Settings.oidc_client_secret
+    )
+    basic_auth_bytes = bytearray(basic_auth_string, 'utf-8')
+    headers['Authorization'] = 'Basic {0}'.format(
+        b64encode(basic_auth_bytes).decode('utf-8')
+    )
+
+    token_request = session.post(token_info_url, data=request, headers=headers)
+    if not token_request.ok:
+        return None
+    json = token_request.json()
+    if "active" in json and json["active"] == False:
+        return None
+    return json
+
+
+def validate_scope(required_scopes, token_scopes) -> bool:
```

Comment:
This function is not used except in tests. Is this WIP or redundant function?

Reply:
No it is required. It is referenced in yaml spec:
`   x-scopeValidateFunc: opera.api.controllers.security_controller.validate_scope `
get_access_token is WIP though